src/titlecase/ascii.rs

```diff
@@ -0,0 +1,295 @@
+use core::fmt;
+use core::iter::FusedIterator;
+
+use bstr::ByteSlice;
+
+#[derive(Clone)]
+#[must_use = "Titlecase is a Iterator and must be used"]
+pub struct Titlecase<'a> {
+    slice: &'a [u8],
+    first: bool,
+}
+
+impl<'a> fmt::Debug for Titlecase<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Titlecase")
+            .field("slice", &self.slice.as_bstr())
+            .finish()
+    }
+}
+
+impl<'a> From<&'a [u8]> for Titlecase<'a> {
+    fn from(slice: &'a [u8]) -> Self {
+        Self::with_slice(slice)
+    }
+}
+
+impl<'a> Titlecase<'a> {
+    pub const fn with_slice(slice: &'a [u8]) -> Self {
+        Self { slice, first: true }
+    }
+}
+
+impl<'a> Iterator for Titlecase<'a> {
+    type Item = u8;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let (&byte, remainder) = self.slice.split_first()?;
+        self.slice = remainder;
+        if self.first {
+            self.first = false;
+            Some(byte.to_ascii_uppercase())
+        } else {
+            Some(byte.to_ascii_lowercase())
+        }
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let len = self.slice.len();
+        (len, Some(len))
+    }
+
+    fn count(self) -> usize {
+        self.slice.len()
+    }
+}
+
+impl<'a> DoubleEndedIterator for Titlecase<'a> {
+    fn next_back(&mut self) -> Option<Self::Item> {
+        let (&byte, remainder) = self.slice.split_last()?;
+        self.slice = remainder;
+        if remainder.is_empty() {
```

Comment:
There is a bug here with this sequence:

1. Create iterator
2. Call iter.next() to consume the title case / first case.
3. Call iter.next_back() until the iterator is consumed.

This will result in 2 uppercase ASCII chars being emitted.

Reply:
Thanks! Reproduced in 42e1f79 and fixed in 6ffc252.